src/libs/safe-path/src/pinned_path_buf.rs

```diff
@@ -0,0 +1,380 @@
+// Copyright (c) 2022 Alibaba Cloud
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use std::ffi::CString;
+use std::fs::{self, File, Metadata, OpenOptions};
+use std::io::{Error, ErrorKind, Result};
+use std::ops::Deref;
+use std::os::unix::fs::OpenOptionsExt;
+use std::os::unix::io::{AsRawFd, FromRawFd, RawFd};
+use std::path::{Path, PathBuf};
+
+use crate::scoped_join;
+
+/// A pinned version of `PathBuf` to protect from TOCTTOU style of attacks.
+///
+/// A `PinnedPathBuf` is a resolved path buffer with a pinned filesystem object, which guarantees:
+/// - the filesystem object referenced by `PinnedPathBuf::as_path()` never changes.
+/// - the value of `PinnedPathBuf::as_path()` never changes.
+/// - the value of [PinnedPathBuf::target()] is a resolved path, which should never be a
+///   symbolic link, but the target referenced by it may change.
+///
+/// Note:
+/// - Though the value of [PinnedPathBuf::target()] never changes, but the filesystem object
+///   referenced by it may be changed. Only the filesystem object referenced by
+///   `PinnedPathBuf::as_path()` is stable.
+/// - Once the `PinnedPathBuf` object gets dropped, the Path returned by `PinnedPathBuf::as_path()`
+///   gets invalid.
+///
+/// With normal [PathBuf], there's a race window for attackers between time to validate a path and
+/// time to use the path. An attacker may maliciously change filesystem object referenced by the
+/// path by symlinks to compose an attack.
+///
+/// The [PinnedPathBuf] is introduced to protect from such attacks, by using the `/proc/self/fd/xxx`
```

Comment:
It used to be the case that a `/proc/` mount was not required to run a Linux system. That situation might have changed (systemd appears to require it), but it might be worth noting this crate requires it.

Reply:
`/proc/` has been required for decades in practice, if not in theory.